### PR TITLE
Multispeaker bugfix and  VITS multispeaker d-vector training support

### PR DESCRIPTION
**Pull request by Edresson**

### CODE REVIEW

File: TTS/trainer.py [hunk 1]
```diff
@@ -210,10 +210,21 @@ def __init__(
...
-
         # init multispeaker settings of the model
         if hasattr(self.model, "init_multispeaker"):
             self.model.init_multispeaker(self.config, self.data_train + self.data_eval)
+            config = config.model_args if hasattr(self.config, "model_args") else self.config
+            # save speakers json
+            if config.use_speaker_embedding and not config.use_d_vector_file:
+                self.model.speaker_manager.save_speaker_ids_to_file(os.path.join(self.output_path, "speaker_ids.json"))
+            elif config.use_speaker_embedding and config.use_d_vector_file:
+                self.model.speaker_manager.save_d_vectors_to_file(os.path.join(self.output_path, "speakers.json"))
```
Comment (Edresson) [inline TTS/trainer.py:221]: It saves the speaker embeddings/speaker ids in the model's logs folder, thus facilitating reproducibility and avoiding problems.

File: TTS/trainer.py [hunk 2]
```diff
@@ -210,10 +210,21 @@ def __init__(
...
+            if hasattr(self.config, "model_args"):
+                self.config.model_args["num_speakers"] = self.model.num_speakers
+            else:
+                self.config.num_speakers = self.model.num_speakers
+            # update config file
+            copy_model_files(self.config, self.output_path, None)
```
Comment (Edresson) [inline TTS/trainer.py:227]: It saves the number of speakers in the training config, this way, it is no longer necessary to manually set the value to use in the inference.

File: TTS/trainer.py [hunk 3]
```diff
@@ -484,14 +495,14 @@ def _optimize(
                 with amp.scale_loss(loss_dict["loss"], optimizer) as scaled_loss:
                     scaled_loss.backward()
                 grad_norm = torch.nn.utils.clip_grad_norm_(
-                    amp.master_params(optimizer), grad_clip, error_if_nonfinite=False
```
Comment (Edresson) [inline TTS/trainer.py:487]: This parameter is not present in previous pytorch versions and the default is False. To maintain compatibility with pytorch previous versions I removed it.

File: TTS/trainer.py [hunk 4]
```diff
@@ -819,9 +830,12 @@ def test_run(self) -> None:
...
         if self.restore_step != 0 or self.args.best_path:
-            print(" > Restoring best loss from " f"{os.path.basename(self.args.best_path)} ...")
-            self.best_loss = load_fsspec(self.args.restore_path, map_location="cpu")["model_loss"]
-            print(f" > Starting with loaded last best loss {self.best_loss}.")
+            model_state = load_fsspec(self.args.restore_path, map_location="cpu")
+            # ignore the restoring of the best loss if do not exist the key "model_loss" in the model state
+            if "model_loss" in model_state.keys():
+                print(" > Restoring best loss from " f"{os.path.basename(self.args.best_path)} ...")
+                self.best_loss = model_state["model_loss"]
+                print(f" > Starting with loaded last best loss {self.best_loss}.")
```
Comment (Edresson) [inline TTS/trainer.py:838]: This is to allow the loading and continuation of training even if the checkpoint does not have the ["model_loss"] key.

File: TTS/tts/datasets/__init__.py [hunk 5]
```diff
@@ -38,15 +38,16 @@ def load_meta_data(datasets, eval_split=True):
...
         preprocessor = _get_preprocessor_by_name(name)
         # load train set
-        meta_data_train = preprocessor(root_path, meta_file_train)
+        meta_data_train = preprocessor(root_path, meta_file_train, ununsed_speakers=ununsed_speakers)
         print(f" | > Found {len(meta_data_train)} files in {Path(root_path).resolve()}")
         # load evaluation split if set
         if eval_split:
             if meta_file_val:
-                meta_data_eval = preprocessor(root_path, meta_file_val)
+                meta_data_eval = preprocessor(root_path, meta_file_val, ununsed_speakers=ununsed_speakers)
```
Comment (Edresson) [inline TTS/tts/datasets/__init__.py:50]: This implements the ununsed_speakers that previously didn't work. I supported all multispeaker datasets.

File: TTS/tts/models/vits.py [hunk 6]
```diff
@@ -658,8 +667,13 @@ def test_run(self, ap) -> Tuple[Dict, Dict]:
         test_audios = {}
         test_figures = {}
         test_sentences = self.config.test_sentences
-        aux_inputs = self.get_aux_input()
+        if hasattr(self, "speaker_manager"):
+            aux_inputs = self.speaker_manager.get_random_speaker_aux_input()
+        else:
+            aux_inputs = self.get_aux_input()
+
```
Comment (Edresson) [inline TTS/tts/models/vits.py:674]: To synthesize test sentences during multispeaker model training it is necessary to obtain the speaker_ids or d_vecs of some speaker. From this model I implemented a function in speaker_manager that returns a randomly chosen speaker (as was done in the past before the trainer api)

File: TTS/tts/utils/speakers.py [hunk 7]
```diff
@@ -155,15 +155,21 @@ def save_d_vectors_to_file(self, file_path: str) -> None:
...
-        speakers = sorted({x["name"] for x in self.d_vectors.values()})
-        self.speaker_ids = {name: i for i, name in enumerate(speakers)}
+
+        # load speakers from data, because during the training we can just use some speakers from d_vector_file
+        if data is not None:
+            self.speaker_ids, _ = self.parse_speakers_from_data(data)
+        else:
+            speakers = sorted({x["name"] for x in self.d_vectors.values()})
+            self.speaker_ids = {name: i for i, name in enumerate(speakers)}
+
```
Comment (Edresson) [inline TTS/tts/utils/speakers.py:172]: For training with d_vecs, speaker_manager listed all speakers present in speakers.json. But many times we only want to use some talkers (like when we use ununsed_speakers to ignore talkers). To ensure that during training only the speakers that will be used in training are printed, I modified this function.

File: TTS/tts/utils/speakers.py [hunk 8]
```diff
@@ -210,6 +216,14 @@ def get_mean_d_vector(self, speaker_idx: str, num_samples: int = None, randomize
...
         return d_vectors
 
+    def get_random_speaker_aux_input(self) -> Dict:
+        if self.d_vectors:
+            return {"speaker_id": None, "style_wav": None, "d_vector":  self.d_vectors[random.choices(list(self.d_vectors.keys()))[0]]["embedding"]}
+        elif self.speaker_ids:
+            return {"speaker_id": self.speaker_ids[random.choices(list(self.speaker_ids.keys()))[0]], "style_wav": None, "d_vector": None}
+        else:
+            return {"speaker_id": None, "style_wav": None, "d_vector": None}
+
```
Comment (Edresson) [inline TTS/tts/utils/speakers.py:226]: method to randomly choose a speaker to be used in synthesizing test sentences during training.

File: TTS/utils/manage.py [hunk 9]
```diff
@@ -131,11 +133,25 @@ def download_model(self, model_name):
...
+                config = load_config(config_path)
+                # models with model_args
+                if "model_args" in config.keys():
+                    config.model_args["speakers_file"] = output_speakers_id_path
+                else:
+                    # other models
+                    config.speakers_file = output_speakers_id_path
+                config.save_json(config_path)
+
+
```
Comment (Edresson) [inline TTS/utils/manage.py:154]: Fix the issue #721

File: TTS/tts/utils/speakers.py [hunk 10]
```diff
@@ -210,6 +216,15 @@ def get_mean_d_vector(self, speaker_idx: str, num_samples: int = None, randomize
...
 
+    def get_random_speaker_aux_input(self) -> Dict:
+        if self.d_vectors:
+            return {"speaker_id": None, "style_wav": None, "d_vector":  self.d_vectors[random.choices(list(self.d_vectors.keys()))[0]]["embedding"]}
+
+        if self.speaker_ids:
+            return {"speaker_id": self.speaker_ids[random.choices(list(self.speaker_ids.keys()))[0]], "style_wav": None, "d_vector": None}
+
+        return {"speaker_id": None, "style_wav": None, "d_vector": None}
+
```
Comment (Edresson) [inline TTS/tts/utils/speakers.py:227]: method to randomly choose a speaker to be used in synthesizing test sentences during training.

File: TTS/tts/utils/speakers.py (hunk 7 above)
Comment (skol101) [reply TTS/tts/utils/speakers.py:172]: @Edresson if I'm fine-tuning a pre-trained VCTK model on a single new speaker, does this setting (speaker_ids) apply ?

Comment (Edresson) [reply TTS/tts/utils/speakers.py:172]: In this case, your new speaker will have samples in the training so this part of the code will not influence your training. This just counts the number of speakers and returns all speakers that are in the training dataset.

**Pull request closed without merging**